proposals/0046-optimistic-cluster-restart-automation.md

```diff
@@ -0,0 +1,321 @@
+voted fork to all others in restart.
+
+We use a new Gossip message `LastVotedForkSlots`, its fields are:
+
+* `last_voted_slot`: `u64` the slot last voted, this also serves as last_slot
+for the bit vector.
+* `last_voted_hash`: `Hash` the bank hash of the slot last voted slot.
+* `ancestors`: `BitVec<u8>` compressed bit vector representing the slots on
+sender's last voted fork. the most significant bit is always
+`last_voted_slot`, least significant bit is `last_voted_slot-81000`.
+
+The number of ancestor slots sent is hard coded at 81000, because that's
+400ms * 81000 = 9 hours, we assume most restart decisions to be made in 9 
+hours. If a validator restarts after 9 hours past the outage, it cannot join 
+the restart this way. If enough validators failed to restart within 9 hours, 
+then fallback to the manual, interactive cluster restart method.
+
+When a validator enters restart, it uses silent repair shred version to avoid
+interfering with those outside the restart. There is slight chance that 
+the silent repair shred version would collide with the shred version after the
+"silent repair phase", but even if this rare case occurred, we plan to flush 
+gossip on successful restart before entering normal validator operation.
+
+To be extra cautious, we will also filter out `LastVotedForkSlots` and
+`HeaviestFork` in gossip if a validator is not in "silent repair phase".
+
+`LastVotedForkSlots` message will be written into gossip and then be
+distributed to all others in restart. Meanwhile the validator can immediately
+enter next step.
+
+### 2. "silent repair phase": Repair ledgers up to the restart slot
+
+The main goal of this step is to repair all blocks which could potentially be
+optimistically confirmed.
+
+We need to prevent false negative at all costs, because we can't rollback an 
+optimistically confirmed block. However, false positive is okay. Because when 
+we select the heaviest fork in the next step, we should see all the potential 
+candidates for optimistically confirmed slots, there we can count the votes and
+remove some false positive cases.
+
+However, it's also overkill to repair every block presented by others. When
+`LastVotedForkSlots` messages are being received and aggregated, a validator
+can categorize blocks missing locally into 3 categories: ignored, must-have, 
+and unsure. Depending on the stakes of validators currently in restart, some
+slots with too few stake can be safely ignored, some have enough stake they
+should definitely be repaired, and the rest would be undecided pending more
+confirmations.
+
+Assume `RESTART_STAKE_THRESHOLD` is 80% and that 5% restarted validators can
+change their votes from what they voted before the restart due to mistakes or 
+malicious behavior.
+
+When only 5% validators are in restart, everything is in "unsure" category.
+
+When 67% validators are in restart, any slot with less than
+67% - 5% - (100-67%) = 29% is in "ignored" category, because even if all 
```

Comment:
A more detailed breakdown of this math would be helpful to readers, for instance breaking down:

1. 100-67% is the remaining validators that can join
2. 67% is the threshold for OC
3. 5% is the size of the possible equivocating set

Then, even if all 33% of remaining validators joined, this slot could have at most 29 + 33 = 62% of the stake voting on it, so the slot couldn't have been optimistically confirmed

Also then maybe clarifying how the malicous 5% affects this threshold. For instance why not make this ignored threshold 67 - 33 = 34% instead of 29%? What can the malicous actors do? 

Seems like the more intuitive formula would be defining `OVERLAP_SET = (RESTART_STAKE_THRESHOLD + OPTIMISTIC_CONFIRMED_THRESHOLD) - 100`,  which represents the overlap between the optimistic set and the set of restarting validators. Then of the `OVERLAP_SET - MALICIOUS_SET` to get the final threshold, so in this case, `OVERLAP_SET = 67 + 67 - 100 = 33`, subtracting the malicous set would be then `5%`, because the malicious set could have voted towards OC before the restart, but then not voted on the restart

Reply:
Thought some more about this logic, I realized we only need two categories: must-have and ignored. I reworked the paragraph, see if it's now better?